mam/src/mam.rs

```diff
@@ -39,116 +37,148 @@ where
 pub fn create<C, CB, H>(
     seed: &[Trit],
     message: &[Trit],
+    side_key: &[Trit],
+    root: &[Trit],
+    siblings: &[Trit],
+    next: &[Trit],
     start: usize,
-    count: usize,
     index: usize,
-    next_start: usize,
-    next_count: usize,
     security: u8,
-    curl1: &mut C,
-    curl2: &mut C,
+    sponge: &mut C,
     bcurl: &mut CB,
-) -> (Vec<Trit>, Vec<Trit>)
+) -> Vec<Trit>
 where
     C: Curl<Trit>,
     CB: Curl<BCTrit>,
     H: HammingNonce<Trit>,
 {
     // generate the key and the get the merkle tree hashes
-    let mut key = [0 as Trit; iss::KEY_LENGTH];
-    let (siblings, root) = {
-        let mut digest = [0 as Trit; iss::DIGEST_LENGTH];
-
-        let addresses: Vec<Vec<Trit>> = {
-            let mut addr = Vec::new();
-            for i in start..(start + count) {
-                merkle::key(seed, i, security, &mut key, curl1);
-                curl1.reset();
-
-                iss::digest_key::<Trit, C>(&key, &mut digest, curl1, curl2);
-                curl1.reset();
-                curl2.reset();
-                iss::address::<Trit, C>(&mut digest, curl1);
-                curl1.reset();
-                addr.push(digest[..iss::ADDRESS_LENGTH].to_vec());
-            }
-
-            curl1.reset();
+    let mut key: Vec<Trit> = vec![0 as Trit; security as usize * iss::KEY_LENGTH];
+    let mut curl_bak: [Trit; STATE_LENGTH] = [0; STATE_LENGTH];
+    sponge.reset();
 
-            addr
+    merkle::key(seed, start + index, security, &mut key, sponge);
+    sponge.reset();
+    message_key(side_key, root, index, sponge);
+    let mut payload: Vec<Trit> = Vec::new();
```

Comment:
Could we alloc `with_capacity` here and do most of the length caclulations before we get to the nitty gritty?